### DISCUSSION TITLE
Error with golang on provided.al2

### DISCUSSION THREAD

**Discussion by harryzcy** [serverless/serverless#11667] [Q&A]
### Are you certain it's a bug?

- [X] Yes, it looks like a bug

### Is the issue caused by a plugin?

- [X] It is not a plugin issue

### Are you using the latest v3 release?

- [X] Yes, I'm using the latest v3 release

### Is there an existing issue for this?

- [X] I have searched existing issues, it hasn't been reported yet

### Issue description

I'm getting `Couldn't find valid bootstrap(s): [/var/task/bootstrap /opt/bootstrap]` error when using `provided.al2`. How do I configure the functions exactly?

### Service configuration (serverless.yml) content

```yaml
service: example

frameworkVersion: '3'

provider:
  name: aws
  runtime: provided.al2
  memorySize: 128
  stage: ${opt:stage, 'dev'}
  region: ${opt:region, 'us-west-2'}

package:
  patterns:
    - '!./**'
  individually: true

functions:
  exampleFunction1:
    handler: bin/functions/exampleFunction1
    package:
      patterns:
        - './bin/functions/exampleFunction1'
```

### Command name and used flags

N/A

### Command output

```shell
N/A
```

### Environment information

```shell
Framework Core: 3.25.1
Plugin: 6.2.2
SDK: 4.3.2
```

**Comment by medikoo** [comment]
Moved to discussions, as it's more a usage support enquiry and not a bug report.

**Comment by G-OD** [comment]
You need to have a bootstrap binary in order for this to work. The easiest method would be to use the serverless-go-plugin. With that you can do something like this:

```yml
service: example
frameworkVersion: '3'

provider:
  name: aws
  runtime: provided.al2
  memorySize: 128
  stage: ${opt:stage, 'dev'}
  region: ${opt:region, 'us-west-2'}

plugins:
  - serverless-go-plugin

custom:
  go:
    supportedRuntimes: ["provided.al2"]
    buildProvidedRuntimeAsBootstrap: true

functions:
  exampleFunction1:
    handler: functions/exampleFunction1/main.go
```

**Comment by ivaylopivanov** [comment]
@G-OD following the example above I managed to deploy but when I execute the lambda I get:

```
RequestId: 91a7b4f0-4e77-4418-b920-f3d205f199ab Error: Couldn't find valid bootstrap(s): [/var/task/bootstrap /opt/bootstrap]
Runtime.InvalidEntrypoint
```

As mentioned [here](https://github.com/mthenw/serverless-go-plugin/issues/41#issuecomment-1725000434) I noticed that the `.bin` folder is now empty.

**Reply by G-OD** [reply]
This is because you need to compile your binary with the name "bootstrap". I have created a basic demo over at https://github.com/G-OD/Go-Serverless-Example. In production you may not want to proxy all endpoints like this (maybe compile a different binary per endpoint like the serverless aws-go-mod example) but it should give a basic idea on how to set it up.

**Reply by ivaylopivanov** [reply]
Thank you! Yes, I got it that it has to be named `bootstrap` but how would that work when we have multiple functions? Can I please ask you to expand your example by adding 1 more function to the serverless file?

In addition, I see you compile the function separately instead of relying on the go plugin to compile it for you, is this on purpose?

**Reply by G-OD** [reply]
Yes in this example I do not use the Go plugin. If you don't mind proxying all endpoints like this you can just continue to add routes and it'll automatically work. If you want to separate the endpoints then one solution is to point them all to the bootstrap binary. As for compiling them into different binaries I'd have to look into it as the deprecation of the Go runtime seems to of changed a lot. I've updated the repo to reflect this.

I'll be sure to post here once I figure out how to do this as it's rather annoying that Amazon have deprecated this.

TLDR: If you don't mind proxying then just leave the serverless.yml as is, otherwise specify all endpoints and point to the binary as the handler.

**Reply by ivaylopivanov** [reply]
Thank you! Indeed thats one way to do it but that requires internal routing like the one you have implemented. I currently have many individual lambdas that do not have internal routing and need multiple individual `bootstrap` out files, 1 for each lambda. It seems thats a limitation of the serverless framework as of now and I cannot achieve that with it.

**Reply by KarnerTh** [reply]
Had the same issue - have a look at https://blog.matthiasbruns.com/running-multiple-golang-aws-lambda-functions-on-arm64-with-serverlesscom

**Reply by ivaylopivanov** [reply]
this is now fixed in the golang plugin, more info here https://github.com/mthenw/serverless-go-plugin/issues/41
